src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1524,6 +1531,11 @@ public SolutionCompilationState UpdateSpecificSourceGeneratorExecutionVersions(
                 if (newTracker != existingTracker)
                     newIdToTrackerMapBuilder[projectId] = newTracker;
             }
+
+            // Clear out the cache of any previously initialized GeneratorDriver. Otherwise we might reuse a
+            // driver which will not count as a new "run" in some of our unit tests. We have tests that very explicitly count
+            // and assert the number of invocations of a generator.
+            GeneratorDriverCache.EmptyCacheForProject(projectId);
```

Comment:
if EmptyCacheForProject is only for use in testing, can you expose it with a TestAccessor.  I was questioning if it needs to be public when i was looking at it.

Reply:
I don't like this either; this is right now here to get @chsienki unblocked while we dig into this further. I think updating the tests is the right thing to do long term, but the tests are pretty hard to figure out and will need further investigation and cleanup. Right now this errs on the side of not breaking product behavior at the limitation that we're potentially tossing the cache more than we need to.

(See the internal Teams chat on this point specifically.)